packages/@headlessui-react/src/hooks/use-outside-click.ts

```diff
@@ -1,132 +1,131 @@
-import { useEffect, useRef, type MutableRefObject } from 'react'
+import { useCallback, useRef, type MutableRefObject } from 'react'
 import { FocusableMode, isFocusableElement } from '../utils/focus-management'
 import { isMobile } from '../utils/platform'
 import { useDocumentEvent } from './use-document-event'
 import { useIsTopLayer } from './use-is-top-layer'
+import { useLatestValue } from './use-latest-value'
 import { useWindowEvent } from './use-window-event'
 
 type Container = MutableRefObject<HTMLElement | null> | HTMLElement | null
 type ContainerCollection = Container[] | Set<Container>
 type ContainerInput = Container | ContainerCollection
 
+// If the user moves their finger by ${MOVE_THRESHOLD_PX} pixels or more, we'll
+// assume that they are scrolling and not clicking. This will prevent the click
+// from being triggered when the user is scrolling.
+//
+// This also allows you to "cancel" the click by moving your finger more than
+// the threshold in pixels in any direction.
+const MOVE_THRESHOLD_PX = 30
+
 export function useOutsideClick(
   enabled: boolean,
   containers: ContainerInput | (() => ContainerInput),
   cb: (event: MouseEvent | PointerEvent | FocusEvent | TouchEvent, target: HTMLElement) => void
 ) {
   let isTopLayer = useIsTopLayer(enabled, 'outside-click')
+  let cbRef = useLatestValue(cb)
 
-  // TODO: remove this once the React bug has been fixed: https://github.com/facebook/react/issues/24657
-  let enabledRef = useRef(false)
-  useEffect(
-    process.env.NODE_ENV === 'test'
-      ? () => {
-          enabledRef.current = isTopLayer
-        }
-      : () => {
-          requestAnimationFrame(() => {
-            enabledRef.current = isTopLayer
-          })
-        },
-    [isTopLayer]
-  )
```

Comment:
The linked issue was expected behavior, since then we've also restructured some of this logic and this doesn't happen anymore.

The issue we were seeing was that opening a `<Dialog />` by setting state in an `onClick`, registered a `click` listener on the `window` and immediately triggered it so the `<Dialog />` was immediately closed.